synapse/handlers/message.py

```diff
@@ -956,7 +1007,13 @@ async def handle_new_client_event(
             logger.exception("Failed to encode content: %r", event.content)
             raise
 
-        await self.action_generator.handle_push_actions_for_event(event, context)
+        # Skip push notification actions for historical messages
+        # because we don't want to notify people about old history back in time.
+        # The historical messages also do not have the proper `context.current_state_ids`
+        # and `state_groups` because they have `prev_events` that aren't persisted yet
+        # (historical messages persisted in reverse-chronological order).
+        if event.content.get(EventContentFields.MSC2716_HISTORICAL, None) is None:
+            await self.action_generator.handle_push_actions_for_event(event, context)
```

Comment:
@erikjohnston Calling out a new piece of logic here ^

It makes sense to skip push notifications for historical messages but the real reason I did this is just because it throws errors. That logic uses `context.current_state_ids` and the historical messages do not have the proper `context.current_state_ids` and `state_groups` because they have `prev_events` that aren't persisted yet (historical messages persisted in reverse-chronological order).

Reply:
Yeah, I think this makes sense. It'd be good if we had another way of figuring out whether an event is historical (c.f. the comment below)